### PR TITLE
Recaptcha

### PR DESCRIPTION
**Pull request by davidedelvento**
Adding a server configuration to request reCAPTCHA on account creation, useful to safeguard against some attacks.
Important also for #145 when each registration sends and email.

### CODE REVIEW

**Comment by davidedelvento** [comment]
Note that merging pr #145 pr #146 and pr #148 together will confuse git since there are some changes appearing on the same line. This is annoying, but there is no actual conflict, the changes are totally orthogonal and should be simply included from each one of them. See my main branch to see how I merged them together.

**Comment by davidedelvento** [comment]
@lambdaTotoro (sorry for repeating the same text, but since the PR is separate...)

Is there any plan to merge this PR? No pressure, just asking to plan accordingly for some deployments (i.e. if you plan to merge "soon", we'll pull the main repo which we'd love to, otherwise we'll pull my fork).

Let me know if there is anything in my code that you'd like to see changed to ease the merge.

**Comment by lambdaTotoro** [comment]
I think this is a generally sensible way to go and offers a good bit of 'enterprise ready-ness'. 
I'm inclined to merge after I test it (I will probably have time sometime this week). But first: there are multiple versions of reCaptcha. Does this work for all of them? If so / If not, that should be included in the documentation.

**Comment by davidedelvento** [comment]
I'll double check the versions or reCaptcha that are supported and include that in the documentation. Hopefully by the end of today.

**Comment by davidedelvento** [comment]
Ok, so this now should be easier to test since I merged with upstream resolving the fake conflicts arising by just different codes happening on the same line of the source (they were both independent and should have just been concatenated, but git did not recognize that). 

I also added a few links and clarified the documentation. Looking forward to seeing if it works for you and happy to work more on it if necessary.

File: docs/options.rst [hunk 1]
```diff
@@ -78,10 +78,20 @@ Native Authenticator is based on username and password only. But if you need ext
...
+
+    c.Authenticator.recaptcha_key = "your key"
+    c.Authenticator.recaptcha_secret = "your secret"
+
+
 Force acceptance of Terms of Service before SignUp
 --------------------------------------------------
 
 You may force users to click a checkbox agreeing to TOS before they can Sign Up. This might be legally binding is some jurisditions.
+>>>>>>> 010da1be5d8b39b195df36f05a49b11dd9edcb7d
```
Comment (lambdaTotoro) [inline docs/options.rst:94]: This still needs to go

Comment (davidedelvento) [reply docs/options.rst]: Doh! Really sorry about that, not sure how that slipped in! Anyway, I've just removed it and updated the PR.

Thanks for catching it!

**Comment by lambdaTotoro** [comment]
I checked the current state and while it mostly looks good, there's still a problem.

If I enable the feature, it shows _two_ enabled "Create user" buttons on the signup page. I would have expected to only see one that starts as disabled and becomes enabled after passing the captcha (and/or clicking accept on TOS). Alternatively, just erroring on click while not having passed the captcha would also be fine.

double_button_none_disabled

I think this can be resolved with another pass over `signup.html` while sufficiently caffeinated. 😉

**Comment by davidedelvento** [comment]
Apologizes about that too. I got caffeinated enough, doubled checked everything and fixed the problem.
FWIW, I suspect the button was not recognized to be the same button when (I or git itself) merged between TOS and reCAPTCHA.

Thanks for your time and patience with these issues.

**Comment by lambdaTotoro** [comment]
Okay, this seems fine now. Merging!

**Pull request merged**